### DISCUSSION TITLE
Patterns/rules for applying function level instrumentation

### DISCUSSION THREAD

**Discussion by ctron** [guacsec/trustify#756] [Ideas]
We've been applying instrumentation (like ``#[instrument]`) on functions when necessary. And that's fine.

However, I am not sure what the right approach is when it comes to handling fields and return/errors. This also touches a bit on the log levels. Also, I do it wrong myself. So I'd appreciate a guideline I could look a too.

Checking back with syslog (RFC5424) there are severities defined like this:

```
Emergency: system is unusable
Alert: action must be taken immediately
Critical: critical conditions
Error: error conditions
Warning: warning conditions
Notice: normal but significant condition
Informational: informational messages
Debug: debug-level messages
```

I know, we don't have all of them in the log levels, but close enough.

I guess everything of error and above is the same for us. Above error might be a panic on the console. But that's not part of the log system of the application.

When taking a look at what we do, and what `tracing` does by default, then that's spamming a lot of information on the INFO level. Which is ok, if we assume that the application will be run on WARN. Because other than a developer debugging, only WARN and above would be important for an operator of the software.

For instrumenting we mostly have `#[instrument(…, ret)]` or `#[instrument(…, err)]`. Which logs the span with the result on INFO, of the error on ERROR. However, an error on a function isn't necessary a log worthy ERROR for the person operating this software.

So I think we need a set of rules, maybe with examples, that explain how to consistently apply instrumentation and log levels.

**Comment by helio-frota** [comment]
* [Docs link](https://github.com/trustification/trustify/blob/main/docs/design/log_tracing.md)

**Comment by helio-frota** [comment]
And the performance impact:

Random non-scientific examples: 
* `#[instrument]`, 
* `#[instrument(skip_all, ret)]`, 
* `#[instrument(skip_all)]`

![a](https://github.com/user-attachments/assets/b1977b4d-0904-48c8-983b-9459c7d93cbd)
![b](https://github.com/user-attachments/assets/02315e8f-2a48-4429-8ebc-25cdbfdad44b)
![c](https://github.com/user-attachments/assets/a7fee98f-a5b4-4b28-9162-6cb28573f6e7)

**Comment by helio-frota** [comment]
Currently we have these parts instrumented [link](https://gist.github.com/helio-frota/8abbce3da6c058be37ef33f79f3d01e7)

As the tracing crate is a 'general purpose system' to emit events, and OTEL rust crates make use of it , I think the team needs to decide what needs to be instrumented considering:

* The spamming information ( mentioned in the first comment )
* The performance impact
* Who are the interested in looking at the trace results

Example:

Instrument everything by layers?

* Endpoint
  * Business / service layer
    * DB 

Or specific points of interest ? Or a mix of both ?

**Reply by helio-frota** [reply]
`tracing-actix-web` works differently compared with `actix-web-opentelemetry`

When using `actix-web-opentelemetry` we don't need to add `#[instrument]`  like this https://github.com/helio-frota/t-o/blob/main/src/main.rs#L13 , So my previous comment is partially wrong :+1:

**Reply by ctron** [reply]
I assume this is only true for the initial endpoint function? Further on, service functions (etc), still need the instrumentation macro?

**Reply by helio-frota** [reply]
> Further on, service functions (etc), still need the instrumentation macro?

Yes and also respecting/following the declared-log-level :+1: 

* _Upload endpoint example:_

`OTEL_TRACES_SAMPLER_ARG=1 OTEL_EXPORTER_OTLP_ENDPOINT="http://localhost:4317" cargo run --bin trustd api --db-password trustify --devmode --auth-disabled --tracing enabled`

Result:

![2025-01-23_07-05](https://github.com/user-attachments/assets/00c706bb-b1a4-4925-93e8-e6ead071cab4)

***

`RUST_LOG=info OTEL_TRACES_SAMPLER_ARG=1 OTEL_EXPORTER_OTLP_ENDPOINT="http://localhost:4317" cargo run --bin trustd api --db-password trustify --devmode --auth-disabled --tracing enabled`

![2025-01-23_07-09](https://github.com/user-attachments/assets/947863ad-58ad-4922-ae9a-bceb37df0f82)


```rust
/// An async version of [`decompress`].
#[instrument(skip(bytes), fields(bytes_len=bytes.len()), err(level=tracing::Level::INFO))]
pub async fn decompress_async(
    bytes: Bytes,
    content_type: Option<header::ContentType>,
    limit: usize,
) -> Result<Result<Bytes, Error>, JoinError> {
    Handle::current()
        .spawn_blocking(move || decompress(bytes, content_type, limit))
        .await
}
```

**Reply by helio-frota** [reply]
personal-opinion-only-with-a-limited-knowledge-of-both-crates: 

I prefer `tracing-actix-web` [screenshot](https://github.com/helio-frota/t-o/blob/main/6.png) -- ( But `actix-web-opentelemetry` is working already )

* less magic (we need to instrument each functioin-endpoint) respecting the `tracing` crate usage
* I have no idea how it behaves with utopia yet
  * We have configs in trustify code that binds more strongly with `actix-web-opentelemetry` not sure how to remove those right now
* What if I want to instrument only one endpoint ? how to hide all the others with `actix-web-opentelemetry` ?
* `tracing-actix-web` seems to respect end user mental health when dealing with OTEL versions [link](https://github.com/LukeMathWalker/tracing-actix-web/blob/main/Cargo.toml#L28-L87)
* Show the name of the function instead of the endpoint and easy to use manual instrumentation (span creation) based on `tracing` crate [screenshots](https://github.com/helio-frota/t-o?tab=readme-ov-file#t-o)

* I'm not advocating to change the crate
  * I have no idea if that will work
    * If yes then it will be nice and less risky [see commit comment](https://github.com/trustification/trustify/commit/8b25ecb50aabf2d177dc48687523efe670f3c923)
* I have no idea about the performance impact comparing both

**Reply by helio-frota** [reply]
Apparently they have the same "automatic" instrumentation via [middleware](https://github.com/LukeMathWalker/tracing-actix-web/blob/main/examples/opentelemetry/src/main.rs#L62)... not required to add [instrument attribute macro](https://github.com/LukeMathWalker/tracing-actix-web/blob/main/examples/opentelemetry/src/main.rs#L20) ( no idea about the implications of that btw )

* I'm not using the middleware [in my example](https://github.com/martelinho-de-ouro/t-o/blob/main/src/main.rs#L13-L36) 

* I don't know if actix-web-opentelemetry work with both middelware and manual-attribute-macro-instrumentation

**Reply by helio-frota** [reply]
yes it supports middleware usage [link](https://github.com/OutThereLabs/actix-web-opentelemetry/blob/main/examples/server.rs#L68)

**Comment by helio-frota** [comment]
we don't have OTEL-metrics with `tracing-actix-web`, so we can discard this option and continue with `actix-web-opentelemetry` :+1:

**Reply by helio-frota** [reply]
related to #1198
